codec-http2/src/main/java/io/netty/handler/codec/http2/PriorityStreamByteDistributor.java

```diff
@@ -0,0 +1,428 @@
+                childState.allocate(bytesForChild);
+                bytesAllocated += bytesForChild;
+                connectionWindow -= bytesForChild;
+            }
+            int childBytesAllocated = allocateBytesForTree(child, connectionWindow);
+            bytesAllocated += childBytesAllocated;
+            connectionWindow -= childBytesAllocated;
+            return true;
+        }
+    }
+
+    /**
+     * The remote flow control state for a single stream.
+     */
+    private final class PriorityState {
+        final Http2Stream stream;
+        boolean hasFrame;
+        int streamableBytes;
+        int allocated;
+        int unallocatedStreamableBytesForTree;
+
+        PriorityState(Http2Stream stream) {
+            this.stream = stream;
+        }
+
+        /**
+         * Recursively increments the {@link #unallocatedStreamableBytesForTree()} for this branch in
+         * the priority tree starting at the current node.
+         */
+        void unallocatedStreamableBytesForTreeChanged(int delta) {
+            unallocatedStreamableBytesForTree += delta;
+            if (!stream.isRoot()) {
+                state(stream.parent()).unallocatedStreamableBytesForTreeChanged(delta);
+            }
+        }
+
+        void allocate(int bytes) {
+            allocated += bytes;
+
+            // Also artificially reduce the streamable bytes for this tree to give the appearance
+            // that the data has been written. This will be restored before the allocated bytes are
+            // actually written.
+            unallocatedStreamableBytesForTreeChanged(-bytes);
+        }
+
+        /**
+         * Reset the number of bytes that have been allocated to this stream by the priority
+         * algorithm.
+         */
+        void resetAllocated() {
+            allocate(-allocated);
+        }
+
+        void updateStreamableBytes(int newStreamableBytes, boolean hasFrame) {
+            this.hasFrame = hasFrame;
+
+            int delta = newStreamableBytes - streamableBytes;
+            streamableBytes = newStreamableBytes;
+
+            // Update this branch of the priority tree if the streamable bytes have changed for this node.
+            unallocatedStreamableBytesForTreeChanged(delta);
+        }
+
+        void close() {
+            // Unallocate all bytes.
+            resetAllocated();
+
+            // Clear the streamable bytes.
+            updateStreamableBytes(0, false);
+        }
+
+        boolean hasFrame() {
+            return hasFrame;
+        }
+
+        int unallocatedStreamableBytes() {
+            return streamableBytes - allocated;
+        }
+
+        int unallocatedStreamableBytesForTree() {
+            return unallocatedStreamableBytesForTree;
+        }
+    }
+
+    /**
+     * A connection stream visitor that delegates to the user provided visitor.
+     */
+    private class WriteVisitor implements Http2StreamVisitor {
```

Comment:
@nmittler - Are there reasons outside of the scope of this class why this can not be re-entry?  Because we do the allocation first and then write second are there any restrictions on re-entry from an allocation/writer point of view?  If the only limitations are introduced by the state tracked in this class (`writer`, and `error`, `writing`) then:

1.Removing the `writer` from the signature of the `distribute` method means we don't have to carry this as scope through `distribute` calls.
2. Re-throwing error - agreed this is something we should do. However I'm not sure this needs to prevent re-entry. Couldn't we just throw the most recently detected exception and re-set to `null` right before we throw? As long as 1 iteration makes it all the way through and attempts to either write allocated bytes and/or resets the allocated amount are we OK?
3. If 1 and 2 are valid then `writing` is not needed either.